### ISSUE TITLE
Buggy behaviour using orderby without server annotation

### ISSUE THREAD

**Issue by vcgalpin** [links-lang/links#977]
If the `server` annotation is removed from the function `getResult`, the call to `getResult` from the function `handleFormWithCall()` will return `Failure -- the list has not been sorted`, illustrating that `orderby (y.a)` has had no effect.

```
#fun getResult() server {
fun getResult() {
  var unsortedList = [ (a=3),(a=2)];
  var sortedList = [ (a=2),(a=3)];

  var triedToSortList = for (y <- unsortedList) orderby (y.a) [y];

  if (triedToSortList==sortedList)
    {"Success - the list has been sorted"}
  else
    {"Failure -- the list has not been sorted"}
}

fun handleFormWithCall() {
  replaceNode(<div id="result">
      <p>The outcome of calling getResult() from handleFormWithCall() is: {stringToXml(getResult())}</p>
      </div>,
      getNodeById("result"))
}

fun mainPage (_) {
  page
    <html>
    <body>
      <h1>Test function form</h1>
      <p>The outcome of a direct call to getResult() is: {stringToXml(getResult())}</p>
      <form l:onsubmit="{handleFormWithCall()}">
        <button type="submit">Submit!</button>
      </form>
      <div id="result" />
      </body>
    </html>
}

fun main () {
  addRoute("",mainPage);
  servePages()
}

main()

```

**Comment by jamescheney** [comment]
I believe the `orderby` syntactic sugar is desugared to code using `sortByBase` (not `sortBy` as described in the comment at the beginning of `desugarFors.ml`).  If you use `sortBy` instead of `orderby`, i.e. changing the definition of `triedToSortList` to this:
```
var triedToSortList = sortBy(fun (y) {y.z}, unsortedList);
```
 then things seem to work.  

That's a workaround but this is still a bug: if you use this definition using `sortBYBase` instead you get the buggy behavior again.

Since the actual code of `sortBy` and `sortByBase` is identical, this suggest the problem has something to do with either how `unsafe` declarations are handled, or perhaps the fact that `sortByBase` requires the first argument to produce a record rather than a base value, which might mean we get different results when we use whatever order JavaScript provides on records vs. whatever OCaml does.

**Comment by jamescheney** [comment]
Ah right: if we use this definition 
```
var triedToSortList = sortBy(fun (y) {y}, unsortedList);
```
(using `sortBy` but with the identity function instead of projecting `a` out of `y`) then we also experience the buggy behavior, so this ls almost certainly due to Javascript handling ordering among (its representation of) Links records differently from OCaml/Links's generic `<`.  To test this, replace `getResult` with:
```
fun getResult() {

  if ((a=2) < (a=3))
    {"Success - (a=2) < (a=3)"}
  else
    {"Failure -- (a=2) is not < (a=3)"}
}
```

**Comment by jamescheney** [comment]
It seems that `<` in Links is translated to `LINKS.lt` in `jslib.js`.  But I can't find a definition for `LINKS.lt` so have no idea whether it is doing the right thing with records.  Anyone know? @dhil?

**Comment by jamescheney** [comment]
No wait, that's wrong, as discussed in the comment titled "HACK" here: 
https://github.com/links-lang/links/blob/b46e4810e3b11432fb08ebe26b5eb51595d3a8d1/core/irtojs.ml#L325

So the basic problem is we are just using JavaScript's builtin `<`, which is presumably not doing the right thing with records (however they are translated to JavaScript, I assume just as objects.)

**Comment by dhil** [comment]
Yikes! That sounds like a bug to me. We should only use primitive operators on elements of base type, I think. This is a good project for the backend refactoring that I am hoping to work with the new/incoming intern on.

**Comment by dhil** [comment]
I see now that I essentially echoed the code comment above. I should have read the code comment before commenting :-)

**Comment by jamescheney** [comment]
> We should only use primitive operators on elements of base type

I agree in principle but the current approach relies on the ability to compare at non-base types. 

We could simply require the function argument to `sortBy` to return a base type rather than an unconstrained type, and likewise for the inputs to `<`, but for `sortByBase` the result of the first argument is a record, and this is what we want for interfacing with database code.  So with the current design we need at least to do something to ensure that `<` does the right thing with records, or `orderby` will not work.

One right (or at least well explored) way to do this would be to introduce type classes/overloading via qualified types or constraints, then the dictionary passing needed to generate the correct ordering code would be produced in Links before translation to IR and to JS, but that seems like a much heavier lift: we are currently stuck at step 0, which is figure out how to do FreezeML type inference in a constraint-based way that could be extended with qualified/constrained types.

**Comment by dhil** [comment]
I think #1087 might have fixed the JS-side of this issue.

**Comment by jamescheney** [comment]
From refreshing my memory above it sounds like we'd zeroed in on the inconsistent semantics of `<` on client vs. server as the reason for the different behavior.  So, if the example Vashti gave earlier (and variants) now works because `<` has been fixed to be consistent across client and server, I think this can be closed.

**Comment by jamescheney** [comment]
I checked, and it does look like the test cases above (both @vcgalpin 's original example and my variation that just tests `<`) now seem to do the right thing so I think we can declare this fixed.
